runtime/contract_update_validation.go

```diff
@@ -0,0 +1,434 @@
+	}
+
+	return expected.Type.Equal(foundVarSizedType.Type, validator)
+}
+
+func (validator *ContractUpdateValidator) CheckConstantSizedTypeEquality(expected *ast.ConstantSizedType, found ast.Type) error {
+	foundConstSizedType, ok := found.(*ast.ConstantSizedType)
+	if !ok {
+		return validator.getTypeMismatchError(expected, found)
+	}
+
+	// Check size
+	if foundConstSizedType.Size.Value.Cmp(expected.Size.Value) != 0 ||
+		foundConstSizedType.Size.Base != expected.Size.Base {
+		return validator.getTypeMismatchError(expected, found)
+	}
+
+	// Check type
+	return expected.Type.Equal(foundConstSizedType.Type, validator)
+}
+
+func (validator *ContractUpdateValidator) CheckDictionaryTypeEquality(expected *ast.DictionaryType, found ast.Type) error {
+	foundDictionaryType, ok := found.(*ast.DictionaryType)
+	if !ok {
+		return validator.getTypeMismatchError(expected, found)
+	}
+
+	err := expected.KeyType.Equal(foundDictionaryType.KeyType, validator)
+	if err != nil {
+		return err
+	}
+
+	return expected.ValueType.Equal(foundDictionaryType.ValueType, validator)
+}
+
+func (validator *ContractUpdateValidator) CheckRestrictedTypeEquality(expected *ast.RestrictedType, found ast.Type) error {
+	foundRestrictedType, ok := found.(*ast.RestrictedType)
+	if !ok {
+		return validator.getTypeMismatchError(expected, found)
+	}
+
+	err := expected.Type.Equal(foundRestrictedType.Type, validator)
+	if err != nil || len(expected.Restrictions) != len(foundRestrictedType.Restrictions) {
+		return validator.getTypeMismatchError(expected, found)
+	}
+
+	for index, expectedRestriction := range expected.Restrictions {
+		foundRestriction := foundRestrictedType.Restrictions[index]
+		err := expectedRestriction.Equal(foundRestriction, validator)
+		if err != nil {
+			return validator.getTypeMismatchError(expected, found)
+		}
+	}
+
+	return nil
+}
+
+func (validator *ContractUpdateValidator) CheckInstantiationTypeEquality(expected *ast.InstantiationType, found ast.Type) error {
+	foundInstType, ok := found.(*ast.InstantiationType)
+	if !ok {
+		return validator.getTypeMismatchError(expected, found)
+	}
+
+	err := expected.Type.Equal(foundInstType.Type, validator)
+	if err != nil || len(expected.TypeArguments) != len(foundInstType.TypeArguments) {
+		return validator.getTypeMismatchError(expected, found)
+	}
+
+	for index, typeArgs := range expected.TypeArguments {
+		otherTypeArgs := foundInstType.TypeArguments[index]
+		err := typeArgs.Type.Equal(otherTypeArgs.Type, validator)
+		if err != nil {
+			return validator.getTypeMismatchError(expected, found)
+		}
+	}
+
+	return nil
+}
+
+func (validator *ContractUpdateValidator) CheckFunctionTypeEquality(_ *ast.FunctionType, _ ast.Type) error {
+	// Non storable type
+	panic(errors.NewUnreachableError())
+}
+
+func (validator *ContractUpdateValidator) CheckReferenceTypeEquality(_ *ast.ReferenceType, _ ast.Type) error {
+	// Non storable type
+	panic(errors.NewUnreachableError())
```

Comment:
Don't panic but return an error

Reply:
Ideally we shouldn't reach here, because the checker  already validates this, before reaching here.

Anyway updated to return an error to be on the safe side.